lib/tls/session.toit

```diff
@@ -470,43 +468,34 @@ class Session:
   // handshaking is complete.  For this we need to know how many handshake
   // records were sent after encryption was activated.
   flush-outgoing_ -> none:
-    from := 0
-    pending-bytes := #[]
-    while true:
-      fullness := tls-get-outgoing-fullness_ tls_
-      if fullness > from:
-        while fullness - from > bytes-before-next-record-header_:
-          // We have the start of the next record available.
-          if fullness - from + RECORD-HEADER-SIZE_ >= bytes-before-next-record-header_:
-            // We have the full record header available.
-            header := RecordHeader_ outgoing-buffer_[from + bytes-before-next-record-header_..]
-            record-size := header.length
+    // Get the outgoing data from the buffer, freeing up space for more data.
+    outgoing-data := tls-get-outgoing_ tls_
+
+    // Scan the outgoing buffer for record headers.
+    for scan := 0; scan < outgoing-data.size; :
+      if bytes-before-next-record-header_ > 0:
+        skip := min
+            bytes-before-next-record-header_
+            outgoing-data.size - scan
+        scan += skip
+        bytes-before-next-record-header_ -= skip
+      else:
+        addition := min
+            (RECORD-HEADER-SIZE_ - outgoing-partial-header_.size)
+            (outgoing-data.size - scan)
+        if addition != 0:
+          outgoing-partial-header_ += outgoing-data[scan .. scan + addition]
```

Comment:
I don't like reading the `outgoing-partial-header_` field over and over again. Maybe cache it too? It doesn't change here.